build: at least teamcity-test.sh fails to collect artifacts

(I believe this issue exists ever since the scripts were converted/improved to use JSON. I had noticed this already months ago. Wasn't sure before today)

Symptom: my Go unit tests fails after they create temporary artifacts in their $TMPDIR, however the "Artifacts" tabs in TeamCity remains empty.

Example here: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_Test/2029726?buildTab=overview

(In contrast, roachtest failures do get their artifacts populated)

Compare:

- run `make test ... TESTFLAGS=-v` locally, see that the test creates temporary files/directories in $TMPDIR

- same failing test in CI: no artifacts collected

I am surprised because I seem to read in the scripts that TMPDIR is set to a sub-directory of `artifacts` which is mounted in the docker image? 

(But then maybe that directory just exists but does not get copied to the place TeamCity looks for artifacts?)

Is there some expectation that `TMPDIR` gets put into artifacts? I didn't think so and I haven't seen this spelled out anywhere. As far as I can tell, artifacts always go to `./artifacts`. In fact, in `teamcity-*.sh` we usually `export TMPDIR=$PWD/artifacts/x`.

But, that being the case, I shouldn't be seeing this in your test run:

```
--- FAIL: TestUnavailableZip (9.30s)
test_log_scope.go:77: test logs captured to: /tmp/logTestUnavailableZip871213965
```

Clearly `TMPDIR` isn't overridden here. Actually there's even a second level of override in `builder.sh`:

https://github.com/cockroachdb/cockroach/blob/8b0e347176defe948b67b47c8d7d52e95aa732a3/build/builder.sh#L61

So doubly you should not land in `/tmp` here.

> (In contrast, roachtest failures do get their artifacts populated)

They don't use `TMPDIR` but the `--artifacts` variable which defaults to `./artifacts`.

--- a/build/builder.sh
+++ b/build/builder.sh
@@ -173,18 +173,24 @@ set +e
 # -i causes some commands (including `git diff`) to attempt to use
 # a pager, so we override $PAGER to disable.
 
+# NB: for some mysterious reason, TMPDIR gets lost if passed via --env into
+# docker run below, hence the workaround to invoke `env` inside of the
+# container. See:
+#
+# https://github.com/cockroachdb/cockroach/issues/50507
+# https://github.com/benesch/autouseradd/issues/2
+#
 # shellcheck disable=SC2086
 docker run --init --privileged -i ${tty-} --rm \
   -u "$uid:$gid" \
   ${vols} \
   --workdir="/go/src/github.com/cockroachdb/cockroach" \
-  --env="TMPDIR=/go/src/github.com/cockroachdb/cockroach/artifacts" \
   --env="PAGER=cat" \
   --env="GOTRACEBACK=${GOTRACEBACK-all}" \
   --env="TZ=America/New_York" \
   --env=COCKROACH_BUILDER_CCACHE \
   --env=COCKROACH_BUILDER_CCACHE_MAXSIZE \
-  "${image}:${version}" "$@"
+  "${image}:${version}" env TMPDIR=/go/src/github.com/cockroachdb/cockroach/artifacts "$@"
 
 # Build container needs to have at least 4GB of RAM available to compile the project
 # successfully, which is not true in some cases (i.e. Docker for MacOS by default).

